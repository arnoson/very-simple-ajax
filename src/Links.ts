import { copyAttributes } from './utils/copyAttributes'
import { copyNode } from './utils/copyNode'
import { emit } from './utils/emit'
import { merge } from './utils/merge'
import { replaceWith } from './utils/replaceWith'

export interface VisitOptions {
  action?: 'push' | 'replace' | 'none'
  useCache?: boolean
  cacheId?: string
}

const parser = new DOMParser()

export class Links {
  private cachedDocuments: Record<string, Document> = {}
  private previousUrl: string | undefined

  start(watchHistory = true) {
    if (watchHistory) {
      window.addEventListener('popstate', () => this.handlePopstate())
    }
  }

  async visit(
    url: string,
    { action = 'push', useCache, cacheId }: VisitOptions = {}
  ) {
    emit('before-visit', { url })

    this.cache(cacheId ?? this.previousUrl ?? window.location.pathname)
    await this.render(url, useCache)

    if (action === 'replace') {
      history.replaceState(null, '', url)
    } else if (action === 'push') {
      history.pushState(null, '', url)
    }

    emit('visit', { url })
    this.previousUrl = url
  }

  private async render(url: string, useCache = false) {
    const newDocument =
      (useCache && this.cachedDocuments[url]) || (await this.load(url))

    if (!newDocument) return

    document.title = newDocument.title
    merge(document.head, newDocument.head)
    const [container, newContainer] = this.getContainers(document, newDocument)

    // Copy each node with `copyNode()`, so `<script>` will get executed.
    const fragment = new DocumentFragment()
    for (const child of Array.from(newContainer.childNodes)) {
      fragment.appendChild(copyNode(child))
    }

    const permanentElements = document.querySelectorAll('[data-permanent]')

    container.innerHTML = ''
    container.appendChild(fragment)
    copyAttributes(container, newContainer)

    permanentElements.forEach((el) => {
      const copy = document.getElementById(el.id)
      copy && replaceWith(copy, el)
    })
  }

  private async load(url: string): Promise<Document | undefined> {
    try {
      const response = await fetch(url)
      return parser.parseFromString(await response.text(), 'text/html')
    } catch (e) {
      // There was a network error. We reload the page so the user sees the
      // browser's network error page.
      window.location.reload()
    }
  }

  private cache(url: string) {
    this.cachedDocuments[url] = document.cloneNode(true) as Document
  }

  private getContainers(document: Document, newDocument: Document) {
    const selector = newDocument.head.querySelector<HTMLMetaElement>(
      "meta[name='very-simple-links:container']"
    )?.content

    const container = selector && document.querySelector(selector)
    const newContainer = selector && newDocument.querySelector(selector)

    if (newContainer && !container)
      console.warn(
        `Container '${selector}' doesn't exist, swapping body instead.`
      )

    return container && newContainer
      ? [container, newContainer]
      : [document.body, newDocument.body]
  }

  private handlePopstate = async () => {
    const url = window.location.pathname
    await this.visit(url, { action: 'none', useCache: true })
  }
}
