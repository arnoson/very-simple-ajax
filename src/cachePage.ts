export const cache: Record<string, Document> = {}

export const cachePage = (url: string) =>
  (cache[url] = document.cloneNode(true) as Document)
