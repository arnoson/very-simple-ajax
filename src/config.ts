import { Config } from './types'

export const globalConfig: Required<Config> = {
  watchHistory: true,
  merge: 'replace',
}
