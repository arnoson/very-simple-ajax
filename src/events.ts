import { LinksEvent } from './types'

const handlers: Record<string, Function[]> = {}

export const on = (event: LinksEvent, handler: Function) => {
  handlers[event] ??= []
  handlers[event].push(handler)
}

export const off = (event: LinksEvent, handler: Function) => {
  handlers[event]?.splice(handlers[event].indexOf(handler), 1)
}

export const emit = async (event: LinksEvent, payload: { url: string }) => {
  for (const handler of handlers[event] ?? []) {
    await handler(payload)
  }
}
