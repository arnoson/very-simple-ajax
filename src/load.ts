const parser = new DOMParser()
let progress = 0

export const load = async (url: string): Promise<Document | undefined> => {
  try {
    setProgress(0)
    // Only show the progress bar if the page takes more than 500ms to load.
    const progressDelayTimeout = window.setTimeout(
      () => toggleLoading(true),
      500
    )

    const response = await fetch(url, { headers: { 'X-Very-Simple': '1' } })
    const reader = response.body?.getReader()
    const length = parseInt(response.headers.get('Content-Length') ?? '0')
    let receivedBytes = 0

    const stream = new ReadableStream({
      start(controller) {
        const read = async () => {
          const { done, value } = await reader!.read()

          if (done) {
            controller.close()
            return
          }

          receivedBytes += value.length
          setProgress(receivedBytes / length)
          controller.enqueue(value)

          read()
        }
        read()
      },
    })

    const html = await new Response(stream, {
      headers: { 'Content-Type': 'text/html' },
    }).text()

    clearTimeout(progressDelayTimeout)
    setProgress(1)
    toggleLoading(false)

    return parser.parseFromString(html, 'text/html')
  } catch (e) {
    // There was a network error. We reload the page so the user sees the
    // browser's network error page.
    window.location.reload()
  }
}

const setProgress = (value: number) => {
  progress = value
  document.documentElement.style.setProperty(
    '--simple-progress',
    `${value * 100}%`
  )
}

const toggleLoading = (state: boolean) =>
  document.documentElement.toggleAttribute('data-simple-loading', state)
