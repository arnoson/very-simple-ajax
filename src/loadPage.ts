const parser = new DOMParser()

export type LoadPageResponse = {
  document?: Document
  error?: Error
  response?: Response
}

/**
 * Load and parse an html page.
 * @throws
 */
export const loadPage = async (url: string): Promise<LoadPageResponse> => {
  try {
    // Todo: show progress (https://javascript.info/fetch-progress)
    const response = await fetch(url)
    const document = parser.parseFromString(await response.text(), 'text/html')
    return { document, response }
  } catch (error) {
    return { error }
  }
}
