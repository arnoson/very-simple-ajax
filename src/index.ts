import { emit, off, on } from './events'
import { cache, render } from './render'
import { Config, VisitOptions } from './types'

let previousUrl: string | undefined

export default {
  on,
  off,

  start({ watchHistory = true } = {} as Config) {
    if (watchHistory) {
      window.addEventListener('popstate', async () => {
        await this.visit(window.location.pathname, {
          action: 'none',
          useCache: true,
        })
      })
    }
  },

  async visit(
    url: string,
    { action = 'push', useCache, cacheId }: VisitOptions = {}
  ) {
    await emit('before-visit', { url })

    cacheId = cacheId ?? previousUrl ?? window.location.pathname
    cache[cacheId] = document.cloneNode(true) as Document
    await render(url, useCache)

    if (action === 'replace') {
      history.replaceState(null, '', url)
    } else if (action === 'push') {
      history.pushState(null, '', url)
    }

    await emit('visit', { url })
    previousUrl = url
  },
}
