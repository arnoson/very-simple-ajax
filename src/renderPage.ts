import { cache } from './cachePage'
import { loadPage } from './loadPage'
import { merge } from './utils/merge'
import { copyNode } from './utils/copyNode'

export const renderPage = async (url: string) => {
  let dom = cache[url]
  if (!dom) {
    const { document, error } = await loadPage(url)
    if (error) {
      // There was a network error. We reload the page so the user sees the
      // browser's network error page.
      window.location.reload()
      return
    }
    // Document could still be an error page, if the url doesn't exist, but
    // thats actually a good thing as the user will see the servers error page.
    dom = document
  }

  document.title = dom.title
  merge(document.head, dom.head)

  // Copy each node with `copyNode()`, so `<script>` will get executed.
  const fragment = new DocumentFragment()
  for (const child of Array.from(dom.body.childNodes)) {
    fragment.appendChild(copyNode(child))
  }

  // Clone the body to preserve any attributes.
  document.body = dom.body.cloneNode() as HTMLBodyElement
  document.body.appendChild(fragment)
}
