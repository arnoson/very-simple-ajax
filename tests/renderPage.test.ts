import { cache } from '../src/cachePage'
import { renderPage } from '../src/renderPage'
import fs from 'fs/promises'
import { resolve } from 'path'

const loadFile = async (name) =>
  (await fs.readFile(resolve(__dirname, `assets/${name}`))).toString()

describe('render page', () => {
  it('merges the heads and swaps the body', async () => {
    document.documentElement.innerHTML = await loadFile('old-page.html')

    const parser = new DOMParser()
    cache['/url'] = parser.parseFromString(
      await loadFile('new-page.html'),
      'text/html'
    )

    await renderPage('/url')
    expect(document.documentElement).toMatchSnapshot()
  })
})
