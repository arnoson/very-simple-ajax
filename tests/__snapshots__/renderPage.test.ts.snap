// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render page merges the heads and swaps the body 1`] = `
<html>
  <head>
    
    
    <meta
      charset="UTF-8"
    />
    
    
    <meta
      content="IE=edge"
      http-equiv="X-UA-Compatible"
    />
    
    
    <meta
      content="width=device-width, initial-scale=1.0"
      name="viewport"
    />
    
    
    <title>
      New Title
    </title>
    
    
    <script
      src="old-script.js"
    />
    
    
    <script
      defer=""
      src="common-script.js"
    />
    
  
    <script
      src="new-script.js"
    />
  </head>
  
  
  <body
    class="new-body"
  >
    
    
    <h1>
      New Body
    </h1>
    
  


  </body>
</html>
`;
